src/backend_model.cc

```diff
@@ -317,75 +288,140 @@ TritonModel::GetExecutionPolicy(const inference::ModelConfig& model_config)
+    std::string* backend_libdir, std::string* backend_libpath,
+    bool* is_python_based_backend)
+{
+  // Try C++ runtime
+  *backend_libname = AssembleCPPRuntimeLibraryName(backend_name);
+  RETURN_IF_ERROR(FindBackendLibraryPath(
+      search_paths, *backend_libname, backend_libdir, backend_libpath));
+  if (!backend_libpath->empty()) {
+    *is_python_based_backend = false;
     return Status::Success;
   }
-
-  // If not found, then we are processing a python-based backend.
-  // We look for libtriton_python.so in python backend directory
-  // and model.py in provided custom backend's directory
-  std::string python_backend_dir =
-      JoinPath({global_backend_dir, kPythonBackend});
-  bool is_dir;
-  RETURN_IF_ERROR(IsDirectory(python_backend_dir, &is_dir));
-  if (!is_dir) {
-    return Status(
-        Status::Code::INVALID_ARG, "unable to find '" + global_backend_dir +
-                                       "/python', '" + backend_name +
-                                       "' requires python backend to operate.");
-  }
-  search_paths.emplace_back(python_backend_dir);
-  std::string runtime_model_path =
-      JoinPath({global_backend_dir, backend_name, kPythonFilename});
-  bool exists;
-  RETURN_IF_ERROR(FileExists(runtime_model_path, &exists));
-  if (!exists) {
-    return Status(
-        Status::Code::INVALID_ARG,
-        "unable to find '" + backend_libname + "' or '" + backend_name + "/" +
-            kPythonFilename + "' for model '" + model_name + "', in " +
-            JoinPath({global_backend_dir, backend_name}));
+  // Try Python runtime
+  *backend_libname = kPythonFilename;
+  RETURN_IF_ERROR(FindBackendLibraryPath(
+      search_paths, *backend_libname, backend_libdir, backend_libpath));
+  if (!backend_libpath->empty()) {
+    *is_python_based_backend = true;
+    return Status::Success;
   }
+  // Cannot find runtime
+  return Status(
+      Status::Code::INVALID_ARG,
+      "unable to find backend library for backend '" + backend_name +
+          "', try specifying runtime on the model configuration.");
+}
 
-  *python_runtime_modeldir = JoinPath({global_backend_dir, backend_name});
-  std::string python_backend_libname;
-  RETURN_IF_ERROR(BackendConfigurationBackendLibraryName(
-      kPythonBackend, &python_backend_libname));
+Status
+TritonModel::FindBackendLibraryPath(
+    const std::vector<std::string>& search_paths,
+    const std::string& backend_libname, std::string* backend_libdir,
+    std::string* backend_libpath)
+{
+  backend_libpath->clear();
 
-  RETURN_IF_ERROR(LocateBackendLibrary(
-      search_paths, python_backend_libname, backend_libdir, backend_libpath));
+  for (const auto& path : search_paths) {
+    const auto full_path = JoinPath({path, backend_libname});
+    bool exists = false;
+    RETURN_IF_ERROR(FileExists(full_path, &exists));
+    if (exists) {
+      *backend_libdir = path;
+      *backend_libpath = full_path;
+      break;
+    }
+  }
```

Comment:
I think we may need to introduce some canonicalization/absolute pathing here to avoid arbitrary code execution - CC @nnshah1 .

Assume there is some directory mounted at `/mnt` for whatever reason.

One thing to do a quick check, can you provide something like this:
```
# config.pbtxt

# try to escape default search path `/opt/tritonserver/backends`
runtime: "../../../mnt/my_malicious_lib.so"
```

We probably shouldn't allow escaping the expected search directories, regardless of how likely it is in practice.

Reply:
I agree it would probably cover most cases. Something like this may still be possible, not sure:
```
runtime: ".."
```

`<backend_dir>/../libtriton_...so` :sweat_smile: 

But then you could argue that you just add `..` to your excluded substrings.

---

I tried to just check for special characters like this for the model load CVE initially, but I got push back to just fully canonicalize the path we're using and assert that path lives within the expected search paths. This would be a bit more futureproof to account for some unexpected case of symlinks or something like that.

I'll let @nnshah1 share his thoughts.

Reply:
Add `..` to be safe: [Add .. to runtime library name check](https://github.com/triton-inference-server/core/pull/285/commits/1c2c618df891abbb480eff814e7e57369a589c10)

Reply:
@kthui , @rmccorm4 - We should follow the same logic / utilities as in the model load api. 

I'm not in favor of searching for specific strings as that can be error prone and os specific.

Better to canonicalize the backend directory and the final path and ensure it is within the x/backend directory.

This also allows for relative / absolute paths as well if the user wants - but ensures that the final path is correct.